### ISSUE TITLE
[v10] bug: TypeScript not willing to serialize the type of my client

### ISSUE THREAD

**Issue by anthonyshew** [trpc/trpc#2568]
### Provide environment information

  System:
    OS: macOS 12.5
    CPU: (10) arm64 Apple M1 Pro
    Memory: 152.64 MB / 32.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 16.16.0 - ~/.nvm/versions/node/v16.16.0/bin/node
    Yarn: 1.22.19 - ~/.nvm/versions/node/v16.16.0/bin/yarn
    npm: 8.11.0 - ~/.nvm/versions/node/v16.16.0/bin/npm
  Browsers:
    Chrome: 104.0.5112.101
    Safari: 15.6

### Describe the bug

I am seeing an error in my editor saying: `The inferred type of this node exceeds the maximum length the compiler will serialize. An explicit type annotation is needed.` on my `setupTRPC` for my Next.js project.

This is coming after a migration from v9 to v10 where I was hoping to see this problem resolve. However, it seems to have moved from my root router to the `setupTRPC`.

This leads me to believe that the type annotation is getting too long for the TS interpreter in VSCode so it's punting the effort/complaining. However, I'll let the real TS genies figure it out. 😄 

### To reproduce

Providing a minimal reproduction might not be applicable here? This bug occurs because of lots of code, not little code. 😄 

Rather,[ here is a link to my source code](https://github.com/doggo-sports/coregi/tree/111979699addbf6983aebbdc31f92c02e0f5b58a).

You'll find the TS error on [this line](https://github.com/doggo-sports/coregi/blob/111979699addbf6983aebbdc31f92c02e0f5b58a/packages/trpc/nextjs/client/appWrapper.ts#L14).

Removing [these two falsy declaration values](https://github.com/doggo-sports/coregi/blob/111979699addbf6983aebbdc31f92c02e0f5b58a/packages/trpc/tsconfig.json#L7) from tsconfig will reveal the error!

### 👨‍👧‍👦 Contributing

- [ ] 🙋‍♂️ Yes, I'd be down to file a PR fixing this bug!

**Comment by KATT** [comment]
Thanks for having the repo, helps a lot to have a repo to play with! Would you be able to add me as a contributor? Then I can play around with different things as soon as I get the time

**Comment by sachinraja** [comment]
Thanks for creating this! It makes sense as our proxy types are huge to emit. But users shouldn't need to emit types for the client.

If you did need to emit types for the server:
In your current setup, emitting types for the router would work fine since there's no error there. You'd be free to share/consume that in your client packages. And presumably, you'd never need to emit types for your Next app or any other tRPC client.

That being said, we should look into making all our types as small as possible.

**Comment by SamSokolin** [comment]
Also running into this issue but on the line where I'm constructing the reactQueryHooks:

```typescript
import type { TrpcAppRouter } from "@avenueops/api";
import { createReactQueryHooks, TRPCClientError } from "@trpc/react";

// This line throws the exception
export const trpc = createReactQueryHooks<TrpcAppRouter>();
```

**Comment by SamSokolin** [comment]
Ended up finding an answer to this in the tRPC Discord. Looks like this is a v9 problem (which my project is still currently on). Will try upgrading to v10 as suggested and see if the issue goes away! Given that @anthonyshew is already on v10, looks like he may be having a separate issue.

<img width="1403" alt="CleanShot 2022-09-07 at 16 59 12@2x" src="https://user-images.githubusercontent.com/10885502/188978172-2d2a44c8-2a35-4d27-816e-38a51b613c8f.png">

**Comment by KATT** [comment]
Note-to-self / @trpc/trpc-core. Potential actions

- Easy: Add note that you need `declaration: false` in the docs
- Hard: Revisit how we do the proxy API. Would have a bad impact on DX 😨

**Comment by KATT** [comment]
@sachinraja would you agree that we should close this as sort of out-of-scope or should we do some further actions?

**Comment by sachinraja** [comment]
Agree that we can close as out of scope but we should definitely note this in the docs

**Comment by KATT** [comment]
@amolpatel is `declaration: false` a solution for you?

**Comment by KATT** [comment]
I played around a bit with this here and it doesn't look to bad already?

- https://github.com/KATT/trpc-declaration
- https://github.com/KATT/trpc-declaration/blob/next/build/utils/trpc.d.ts
- https://github.com/KATT/trpc-declaration/blob/next/build/server/_app.d.ts

I notice that the Context is duplicated (the `foo: "bar"`-thing)... but unsure how we could unsure that wasn't the case 🤔

**Comment by IgnusG** [comment]
In my case I could not disable the declarations since I needed to have `composite` on. 
Hmmm, not sure what else to do there 😖

**Comment by KATT** [comment]
Is anyone able to create a reproduction of this issue? Would help to diagnose and potentially fixing it

**Comment by SamSokolin** [comment]
Hey @KATT, I work on the same project as @amolpatel. Unfortunately, I don't think `declaration: false` is an option for us because it's required to use [project references](https://www.typescriptlang.org/docs/handbook/project-references.html), which we rely on pretty heavily throughout our monorepo. We can reliably reproduce the issue but the project isn't open source. I'd be happy to hop on a Zoom with you and walk you through where we're running into it.

In the meantime, we ended up using a [split link](https://trpc.io/docs/v10/links#using-a-splitlink-to-control-request-flow) and two separate routers to unblock ourselves, like so:
![CleanShot 2022-09-27 at 14 19 27](https://user-images.githubusercontent.com/10885502/192605687-a6f165fe-6788-4ee2-a3a8-25004ab3815a.png)
Def a little janky b/c it requires us to use multiple tRPC clients throughout the frontend, but works for now!

**Comment by sachinraja** [comment]
Looking into this now.

**Comment by sachinraja** [comment]
So #2890 should help but here's an optimization that you can make:
```ts
// instead of this
export type AppRouter = typeof appRouter

// use this
type AppRouterType = typeof appRouter
export interface AppRouter extends AppRouterType {}
```

this forces TypeScript to emit the named `AppRouter` type for React instead of fully resolving the type again.

**Comment by sachinraja** [comment]
#2890 cuts the declaration size of the server code in half. I'll do a release for that soon. But that, combined with the above comment, should fix this issue.

**Comment by sachinraja** [comment]
Released in `v10.0.0-proxy-beta.13`, please test that and get back to us.

**Comment by SamSokolin** [comment]
@sachinraja Thanks for the quick turnaround here! Will test w/ our project soon and lyk if we run into any issues. Just to clarify, the result of this is reducing the generated type size by a factor of 1/2? If so, does it perhaps make sense to document a recommended workaround when the issue resurfaces (via split links or otherwise)? Generally curious on your thoughts re: dealing w/ large routers in the future!

**Comment by sachinraja** [comment]
@noty-nick
I can reproduce, will fix soon, thanks.

---
@SamSokolin

> Just to clarify, the result of this is reducing the generated type size by a factor of 1/2? If so, does it perhaps make sense to document a recommended workaround when the issue resurfaces (via split links or otherwise)?

Possibly, but ideally we'd deal with the issue if it resurfaces again. Split links are a good solution though, seems similar to #1226

**Comment by KATT** [comment]
I'm trying to reproduce this in #2947 but to no avail. Feel free to jump in and modify the router I've created there (@datagutt @SamSokolin @anthonyshew @amolpatel).

**Comment by anthonyshew** [comment]
Looks like it's been squashed for me!

**Comment by KATT** [comment]
Okay. This will be closed soon as we have an automated test that does this with 700 procedures. 

Please reopen if you encounter these issues again.
